lib/softoken/kbkdf.c

```diff
@@ -0,0 +1,849 @@
+            return CKR_MECHANISM_PARAM_INVALID;
+        }
+
+        return kbkdf_Pipeline(hSession, (CK_SP800_108_KDF_PARAMS *)pMechanism->pParameter, base_key, ret_key, keySize);
+    default:
+        /* Shouldn't happen unless NIST introduces a new KBKDF type. */
+        PR_ASSERT(PR_FALSE);
+    }
+
+    return CKR_MECHANISM_PARAM_INVALID;
+}
+
+CK_RV kbkdf_Counter(CK_SESSION_HANDLE hSession, CK_SP800_108_KDF_PARAMS_PTR params, SFTKObject *base_key, SFTKObject *ret_key, CK_ULONG keySize) {
+    CK_RV ret;
+
+    /* Context for our MAC / PRF function.
+     *
+     * TODO: use PKCS#11 instead of freebl for this?
+     *
+     * Zeroing context required unconditional call of sftk_MAC_Destroy.
+     */
+    SftkMACCtx ctx = {0};
+
+    /* We need two buffers: one for the contents of the PRF output and one
+     * for the temporary contents of the key, if any additional derived key
+     * material spans a PRF block.
+     *
+     * The PRF buffer will be sized according to the desired output length of
+     * the PRF; this is assumed to be its entire length as PKCS#11 v3.0
+     * doesn't allow for the truncation of the PRF output. */
+    uint8_t *output_buffer = NULL;
+
+    /* The key buffer will be created as necessary when the length of the
+     * current key exceeds the remaining PRF output. */
+    size_t key_index = 0;
+    size_t key_buffer_index = 0;
+    uint8_t *key_buffer = NULL;
+
+    /* In extenuating circumstances, while counter should not exceed 32-bits,
+     * the intermediate values required when calculating num_iterations might,
+     * requiring num_iterations to be 64-bits. */
+    uint32_t counter;
+    uint32_t num_iterations;
+
+    /* While NIST specifies a maximum length (in bits) for the counter, they
+     * don't for the maximum length. It is unlikely, but theoretically
+     * possible for output of the PRF to exceed 32 bits while keeping the
+     * counter under 32-bits in length. Thus, use a 64-bit variable for the
+     * maximum output length. */
+    uint64_t output_length;
+
+    /* First validate our passed input parameters against PKCS#11 v3.0
+     * and NIST SP800-108 requirements. */
+    ret = kbkdf_ValidateCounterParameters(params, keySize);
+    if (ret != CKR_OK) {
+        goto finish;
+    }
+
+    /* Initialize the underlying PRF state.
+     *
+     * TODO: Need to figure out how to extract key for use by the PRF in a
+     * manner usable for re-entering PKCS#11. */
+    ret = sftk_MAC_Init(&ctx, params->prfType, base_key);
+    if (ret != CKR_OK) {
+        goto finish;
+    }
+
+    /* Calculate the number of iterations required and the actual output
+     * length. */
+    ret = kbkdf_CalculateIterations(&ctx, CKM_SP800_108_COUNTER_KDF, params, keySize, &num_iterations, &output_length);
+    if (ret != CKR_OK) {
+        goto finish;
+    }
+
+    /* Allocate memory for the PRF output */
+    output_buffer = calloc(ctx.output_size, sizeof(uint8_t));
```

Comment:
OK, I just noticed you are using calloc/malloc. NSS has 'wrapped' these for style and portability reasons (well it's been a while since NSS was ported to a non-POSIX platform, it's still part of the tradition). Should use the PORT_ versions of these functions.

Reply:
Ah right, sorry. :)